### PR TITLE
WIP: code coverage reporting

### PR DESCRIPTION
**Pull request by tylerjw**
This adds code coverage reporting.  Some of .travis.yml has been commented out to speed up travis runs as this is WIP.

### CODE REVIEW

**Comment by tylerjw** [comment]
This PR uses features from my moveit_ci PR.  This will remain WIP until it is passing CI and the moveit_ci change is submitted.

It is currently timing out during the test.  catkin run_tests has an internal timeout of 20min.  I have tried both with and without compiler optimization.

**Comment by rhaschke** [comment]
It's not the case that `run_tests` have their own specific timeout, we just have a global timeout of 2h.
Obviously running the tests with coverage takes (much?) longer than before. Hence, we can only reduce the runtime by reducing the number of tests.
It looks like, currently, all packages are tested, although we should focus on the source packages only, i.e. packages that are part of the source repository. For some reason packages from other repos are not blacklisted anymore. I'm currently looking into this issue.

**Comment by rhaschke** [comment]
I committed a fix to `moveit_ci` master, such that only source-related packages are tested.
Please, rebase your `moveit_ci` branch onto master and try again.

File: moveit_core/CMakeLists.txt [hunk 1]
```diff
@@ -143,6 +143,8 @@ catkin_package(
     OCTOMAP
     urdfdom
     urdfdom_headers
+  CFG_EXTRAS
+    Coverage.cmake
```
Comment (rhaschke) [inline moveit_core/CMakeLists.txt]: Is this line sufficient to include Coverage.cmake in all downstream packages?

Comment (tylerjw) [reply moveit_core/CMakeLists.txt]: Yes.

**Comment by tylerjw** [comment]
> I committed a fix to `moveit_ci` master, such that only source-related packages are tested.
> Please, rebase your `moveit_ci` branch onto master and try again.

Cool, I'll test it now.  Thank you.

**Comment by tylerjw** [comment]
Sadly this is still timing out :(  I wonder if I could get catkin run_tests to run per library somehow.  But even then we are already really close to the travis limit for runtime.  I wonder if there is some way to make this faster or split up the code coverage reporting so divides up what it does over a couple different builds.  One test for moveit_core and one for everything else, I'll see if I can figure out how to do that. 

https://travis-ci.org/ros-planning/moveit/builds/612601429?utm_source=github_status&utm_medium=notification

**Comment by tylerjw** [comment]
Sweet, I got something to work here!

For context though that report is just moveit_core.  I'm testing a new approach that if it works should allow us to split up code coverage tests across a couple travis builds (because running the tests take so much longer) and codecov.io should just merge the report.

**Comment by tylerjw** [comment]
It looks like it is working but I can't verify because master is broken right now.  I rebased on an older master that did pass and force pushed to test this while the current brakeage is worked out.

**Comment by tylerjw** [comment]
Rebased on master as it looks like a fix might have just gotten merged.  Hopefully this works.

**Comment by tylerjw** [comment]
Getting closer, this time `chomp_moveit_test` because it is timing out because the runtime is longer with coverage reporting.  I increased the timeout to see if I can get it to finish.

File: moveit_experimental/moveit_jog_arm/CMakeLists.txt [hunk 2]
```diff
@@ -63,6 +63,13 @@ install(
 
 install(DIRECTORY include/ DESTINATION ${CATKIN_GLOBAL_INCLUDE_DESTINATION})
 
+# to run: catkin_make -DENABLE_COVERAGE_TESTING=ON package_name_coverage
```
Comment (rhaschke) [inline moveit_experimental/moveit_jog_arm/CMakeLists.txt]: Please remove the comment. It doesn't apply (anymore).

File: moveit_experimental/moveit_jog_arm/CMakeLists.txt [hunk 3]
```diff
@@ -63,6 +63,13 @@ install(
...
+if(CATKIN_ENABLE_TESTING AND ENABLE_COVERAGE_TESTING)
```
Comment (rhaschke) [inline moveit_experimental/moveit_jog_arm/CMakeLists.txt]: The block needs to be moved upfront - before any definition of targets.

File: moveit_experimental/moveit_jog_arm/CMakeLists.txt (hunk 2 above)
Comment (tylerjw) [reply moveit_experimental/moveit_jog_arm/CMakeLists.txt]: done

**Comment by tylerjw** [comment]
I've finally had enough trouble with the `codecov.yml` config file that I'm going to start just testing this locally.  Here are the notes I produced for myself on how to do this in case anyone else needs them:

## create workspace
```bash
cd ~/tmp
mkdir travis
cd travis
mkdir -p ws_moveit/src
cd ws_moveit/src
git clone git@github.com:PickNikRobotics/moveit.git -b p/tylerjw/codecov
wstool init .
wstool merge -y moveit/moveit.rosinstall
wstool update
s (skip)
rosdepinstall
cd ../
```

## set enviroment variables
```bash
export ROS_WS=~/tmp/travis/ws_moveit
export REPOSITORY_NAME=moveit
export CI_SOURCE_PATH=$ROS_WS/src/${REPOSITORY_NAME}
export ROS_DISTRO=melodic
export ROS_REPO=ros
export UPSTREAM_WORKSPACE=moveit.rosinstall
export TEST_BLACKLIST="moveit_ros_perception"
export TEST_WHITELIST="moveit_core"
export CXXFLAGS="-Wall -Wextra -Wwrite-strings -Wunreachable-code -Wpointer-arith -Wredundant-decls -Wno-unused-parameter -Wno-unused-but-set-parameter -Wno-unused-function"
export WARNINGS_OK=false
export TEST="code-coverage"
export CMAKE_ARGS="-DCMAKE_BUILD_TYPE=Debug -DENABLE_COVERAGE_TESTING=ON"
```

## build
```bash
catkin config --extend /opt/ros/melodic --install --cmake-args -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS_RELEASE="-O3" $CMAKE_ARGS --
catkin build --summarize
```

## filter_out function
```bash
filter_out() {
  local PATTERN="$1"; shift
  ## convert input lists into newline-separate lists: | tr ' ;,' '\n'
  ## perform filtering: grep -Fvxf
  ## and convert newlines back to spaces: | tr '\n' ' '
  echo "$*" | tr ' ;,' '\n' | grep -Fvxf <(echo "$PATTERN" | tr ' ;,' '\n') | tr '\n' ' '
}
```

## config test build
```bash
catkin config --blacklist $TEST_BLACKLIST
catkin config --whitelist $TEST_WHITELIST
export ALL_PKGS=$(catkin_topological_order $ROS_WS/src --only-names)
export SOURCE_PKGS=$(catkin_topological_order $CI_SOURCE_PATH --only-names)
export BLACKLIST_PKGS=$(filter_out "$SOURCE_PKGS" "$ALL_PKGS")
catkin config --append-args --blacklist $BLACKLIST_PKGS
```

## build tests
```bash
catkin build --summarize --make-args tests
```

## run tests
```bash
catkin build --catkin-make-args run_tests -- --summarize
```

## print test results
```bash
catkin_test_results
```

## upload to codecov
```bash
cd src/moveit
export VCS_BRANCH_NAME=$(git rev-parse --abbrev-ref HEAD)
export VCS_COMMIT_ID=$(git rev-parse HEAD)
cd ../../
travis_run --title "codecov.io report upload" bash <(curl -s https://codecov.io/bash) -s $ROS_WS -R $CI_SOURCE_PATH
```

**Comment by tylerjw** [comment]
It works!  Coverage report should update soon.  The only thing left on this is to go around enabling coverage on the repos outside of moveit_core.

**Comment by rhaschke** [comment]
> I'm going to start just testing this locally.

Just for your info: you can run the Travis test locally, exactly the same way as Travis would do.
See https://github.com/ros-planning/moveit_ci#running-locally-for-testing

> It works! Coverage report should update soon.

I will have a look tomorrow hopefully.

**Comment by tylerjw** [comment]
I know I can run the travis test locally, I just wanted to be able to run that last step (codecov.io script) multiple times without waiting for a full build and without the docker container just going away.

**Comment by tylerjw** [comment]
I know the recent change to how the code-coverage ros package will make most of the changes to the cmake files in this PR irrelevant I wanted to start testing running it over the whole codebase.  I expect to get timeouts on builds as building with coverage takes longer and therefore I'll need to split up the project into multiple travis jobs for code coverage.

**Comment by tylerjw** [comment]
The method used in this project of editing each cmake file seems error prone in that it is a bunch of copied/pasted boilerplate and it would be easy to leave a package out.

**Comment by rhaschke** [comment]
> The method used in this project of editing each cmake file seems error prone in that it is a bunch of copied/pasted boilerplate and it would be easy to leave a package out.

I fully agree. As far as I understand it:
- You need to add compiler flags to instrument the code for coverage reporting. This is done via the `code_coverage` ROS package, but it essentially only adds some compiler options.
- To limit code instrumentation to certain code / packages only, we need to manually edit the cmake files.
- As we instrument full packages, including the test code, you need to explicitly ignore tests for reporting.

If we instrument all the code in the workspace, we could just pass the instrumentation options globally - as you did in the very beginning. This would just mean to extend the ignore tests to all upstream packages. I have seen that you struggled with those ignores. Did you find a way to ignore all stuff within a (package) subfolder? If so, we could simply add all those upstream package folders to the ignore list.

Key question remaining: Why does Travis become so slow?
- Is it due to compiling?
- Is it due to slower test execution (because of building in debug mode w/o optimization?
In this latter case, can we find a better compromise between optimization and good/readable coverage reporting?

File: .travis.yml [hunk 4]
```diff
@@ -21,6 +21,8 @@ env:
     - TEST_BLACKLIST="moveit_ros_perception"
     - CXXFLAGS="-Wall -Wextra -Wwrite-strings -Wunreachable-code -Wpointer-arith -Wredundant-decls -Wno-unused-parameter -Wno-unused-but-set-parameter -Wno-unused-function"
     - WARNINGS_OK=false
+    - CCACHE_SLOPPINESS="include_file_ctime,include_file_mtime"
```
Comment (rhaschke) [inline .travis.yml]: This seems to be dangerous if ccache only looks at the include file's timestamps but not their contents.

Comment (tylerjw) [reply .travis.yml]: This has to do with what it decides to put in the ccache not what it compares with.

From the man page:

include_file_ctime
By default, ccache also will not cache a file if it includes a header whose ctime is too new. This option disables that check.

include_file_mtime
By default, ccache will not cache a file if it includes a header whose mtime is too new. This option disables that check.

**Comment by tylerjw** [comment]
> Key question remaining: Why does Travis become so slow?
> 
> * Is it due to compiling?
> * Is it due to slower test execution (because of building in debug mode w/o optimization?
>   In this latter case, can we find a better compromise between optimization and good/readable coverage reporting?

I think it is actually both.  You will notice I had to add support for splitting up the build based on the whitelist to get under the 60min compile timeout.  The tests themselves also take ~4x the runtime.  I think it has more to do with modifying the coverage reports as they run than building without optimization.

> Did you find a way to ignore all stuff within a (package) subfolder? If so, we could simply add all those upstream package folders to the ignore list.

No, I never got this working.  What I did get working was filtering out sending reporting for everything within a `test` directory recursively.  If we add the flags globally we will be makeing the compile/test runtimes even slower as the upstream packages will now also be built with these flags.  Ideally we would find a way to not build the test targets with the coverage flags making the compiling and running of them faster.  The way I'm doing it now I'm filtering before uploading.  I think my current method is better than the yml file because it is less error prone and the yml file filters on the codecov website making the upload and report generation step longer.

**Comment by tylerjw** [comment]
The last run was successful btw and here is the report:
https://codecov.io/gh/PickNikRobotics/moveit/tree/70809b9ac08d5baf6ddeed521622e79ae3e680d4

**Comment by tylerjw** [comment]
The reason I'm trying to get ccache to work well is I think it could really help improve build times.  Right now build times are really long because we are only getting a ~13% cache hit rate.  I think this is due to the debug symbols.

**Comment by rhaschke** [comment]
> Right now build times are really long because we are only getting a ~13% cache hit rate.

Actually, we had hit rates as high as 92% in the past: https://travis-ci.org/ros-planning/moveit/jobs/613923548. Maybe your build options change for each run?

Could you please try to use optimization for coverage and compare the report results and build/test run times?

**Comment by davetcoleman** [comment]
@tylerjw Is this still WIP?

**Comment by tylerjw** [comment]
The build will fail if I turn on optimization because the `code-coverage` package detects that and throws an error when cmake runs.

We are getting 92% hit rate without debug builds.  I'm going to work on this locally when I have time and see if I can figure out how to get better cache performance.

@davetcoleman yes, still WIP due to really long build times.

**Comment by tylerjw** [comment]
Copying and pasting into all the cmake files is bad.  But the way this project is designed as a bunch of federated projects that seems like the best approach.  My original approach with using an INTERFACE library in the build step with PUBLIC and PRIVATE interfaces to only build coverage into the source libraries and not into the tests seems ideal as building with coverage enabled is slower than without because of the extra files it has to produce.  Right now we are "solving" this by ignoring those coverage reports generated by those test source files when compiling the report.

However the issue of ccache getting cache misses is probably a larger portion of the increased build and run-times and should be solved first.

**Comment by rhaschke** [comment]
> However the issue of ccache getting cache misses is probably a larger portion of the increased build and run-times and should be solved first.

Did you made progress in debugging this?

**Comment by rhaschke** [comment]
Merged simpler solution #1776.

**Pull request closed without merging**